browser/extensions/extensibles/test/browser/browser_privatetab.js

```diff
@@ -27,18 +41,152 @@ add_task(async function testContainer() {
 
 // Test clicking context menu item changes userContextId to match private container
 add_task(async function testTogglePrivateTab() {
-  await BrowserTestUtils.openNewForegroundTab(gBrowser);
-  let tab = gBrowser.selectedTab;
+  let tab = await BrowserTestUtils.openNewForegroundTab(gBrowser);
   is(tab.userContextId, 0, "Tab is not private");
-  await openTabContextMenu(tab);
-  let openPrivate = document.getElementById("toggleTabPrivateState");
-  openPrivate.click();
-  tab = gBrowser.selectedTab;
+  // Toggling private opens a duplicate tab in new container
+  tab = await togglePrivate(tab);
   is(
     tab.userContextId,
     window.PrivateTab.container.userContextId,
     "Tab is private"
   );
+  // Cleanup
+  BrowserTestUtils.removeTab(tab);
+});
+
+// Test autofill not stored from private tab
+add_task(async function testAutofillNotStored() {
+  const URL =
+    "http://mochi.test:8888/browser/browser/components/" +
+    "sessionstore/test/browser_formdata_sample.html";
+
+  const OUTER_VALUE = "browser_formdata_" + Math.random();
+  const INNER_VALUE = "browser_formdata_" + Math.random();
+
+  // Creates a tab, loads a page with some form fields,
+  // modifies their values and closes the tab.
+  async function createAndRemoveTab() {
+    // Create a new tab.
+    let tab = await BrowserTestUtils.openNewForegroundTab(gBrowser, URL);
+    // Toggle to a private tab
+    tab = await togglePrivate(tab, true);
+    let browser = tab.linkedBrowser;
+    await promiseBrowserLoaded(browser);
+
+    // Modify form data.
+    await setPropertyOfFormField(browser, "#txt", "value", OUTER_VALUE);
+    await setPropertyOfFormField(
+      browser.browsingContext.children[0],
+      "#txt",
+      "value",
+      INNER_VALUE
+    );
+
+    // Remove the tab.
+    await promiseRemoveTabAndSessionState(tab);
+  }
+
+  await createAndRemoveTab();
+  let [
+    {
+      state: { formdata },
+    },
+  ] = JSON.parse(SessionStore.getClosedTabData(window));
+  ok(!formdata, "Form data not stored in private tab");
+});
+
+// Test tab history not stored from private tab
+add_task(async function testTabHistoryNotStored() {
+  // Open a new tab
+  let tab = await BrowserTestUtils.openNewForegroundTab(gBrowser, URI1);
+  // Make it private
+  tab = await togglePrivate(tab, true);
+  let browser = tab.linkedBrowser;
+  await promiseBrowserLoaded(browser);
+  // Open a new URL
+  BrowserTestUtils.loadURI(browser, URI2);
+  await promiseBrowserLoaded(browser, false, URI2);
+  // Remove tab to save state
+  BrowserTestUtils.removeTab(tab);
+  // Verify only non-private data stored
+  let closedTabData = JSON.parse(SessionStore.getClosedTabData(window)).filter(
+    data => {
+      return (
+        data.state.entries[0].url == URI1 || data.state.entries[0].url == URI2
+      );
+    }
+  );
+  let privateData = closedTabData.filter(data => {
+    return data.state.isPrivate == true;
+  });
+  is(closedTabData.length, 1, "Tab data not stored in private mode");
```

Comment:
Is this simply saying 'tab data entry exists for the closed private tab', and the next line is saying, 'yeah, but it's empty'?

Reply:
Looking through this the test is maybe a little weird. But my logic was as follows:
1. Open a tab with URI1, toggle it to be private
2. Load URI2 when in private mode so we definitely have some sort of state to save.
3. As toggling private mode duplicates the current tab in the "Private" container and deletes the non-private tab, we should expect `closedTabData` to have one entry, the non-private version of the tab that was remove during toggling. If it has two entries then our private tab state is being stored as well.
4. Private tabs have the `isPrivate` attribute set to true, which should in turn prevent any state being saved. So there should be no state data for tabs with `isPrivate === true`.

Reply:
OK. So my question here is that the two `is()`s have the same comment. I think it might be more intuitive if they were combined, something like:

```
const oneClosedTabWithNoPrivateData = closedTabData.length === 1 && privateData.length === 0;
is(oneClosedTabWithNoPrivateData, true, "Tab data not stored in private mode");
```

or they had different comments as per my interpretation above.

...or we can just leave it if you think it's fine.

WDYT?

Reply:
I like the suggestion you gave - I'll update this now